### PR TITLE
w_string_suffix bugfix and unit test

### PR DESCRIPTION
**Pull request by kaitian521**
1. if a string ends with a '.', absolutely it dose not have suffix
2. `str->len - end >= sizeof(name_buf)` is more readable and comprehensible
3. add unit test for w_string_suffix.

### CODE REVIEW

**Comment by wez** [comment]
Thanks for contributing!

Can you also add a unit test for this function?
We don't have one today, but it should be relatively easy to add one.

Take a look at `tests/argv.cpp` for an example; please add a `tests/string.cpp` file that demonstrates the expected behavior for a few different input strings, including this case where the name ends with a `.`.

You'll also need to edit `Makefile.am` and `winbuild/Makefile` to have the test build and run.

**Comment by kaitian521** [comment]
i should have added some unit test for this function but i failed to find such files... so will i just add a unit test for this func, or i add unit tests for all functions in string.cpp and watchman_string.h? thank you for reply

**Comment by wez** [comment]
Just coverage for the function you're changing is all that is needed here.  If you'd like to add coverage for other string functions, please send a separate pull request!

**Comment by kaitian521** [comment]
@wez, I do not have windows in hand, so let ci run the test, and if something wrong, I will fix it

File: Makefile.am [hunk 1]
```diff
@@ -268,6 +269,13 @@ tests_log_t_SOURCES = \
 	tests/log.cpp \
 	log.cpp
 
+tests_string_t_CPPFLAGS = $(THIRDPARTY_CPPFLAGS) @IRONMANCFLAGS@
+tests_string_t_LDADD = $(JSON_LIB) $(TAP_LIB)
+tests_string_t_SOURCES = \
```
Comment (wez) [inline Makefile.am]: please add `tests/log_stub.cpp` in here; that will allow you to remove the `w_log` stub from the test case

File: tests/string.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/* Copyright 2012-present Facebook, Inc.
+ * Licensed under the Apache License, Version 2.0 */
+
+#include "watchman.h"
+#include <algorithm>
+#include "thirdparty/tap.h"
+
+void w_log(int level, WATCHMAN_FMT_STRING(const char *fmt), ...) {
```
Comment (wez) [inline tests/string.cpp]: please remove this function; there are stubs for this in `log_stub.cpp` that you should use instead

File: tests/string.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  unused_parameter(level);
+  unused_parameter(fmt);
+}
+
+void test_suffix() {
+
+  w_string_t *suffix = nullptr, *str = nullptr, *expect_suffix = nullptr;
+  str = w_string_new_typed("", W_STRING_UNICODE);
```
Comment (wez) [inline tests/string.cpp]: prefer to use the newer smart pointer type for this:

```
w_string str("");
```

Note that it doesn't make too much sense to use `W_STRING_UNICODE` in here.  That is a tag to help the serialization layer recreate string or binary representations as appropriate.

Filenames in watchman are always binary which happens to be the default type when constructing a string via the `w_string` constructor, so you can just omit it.

File: tests/string.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  ok(!suffix, "end with dot");
+
+  str = w_string_new_typed("nosuffix", W_STRING_UNICODE);
+  suffix = w_string_suffix(str);
+  ok(!suffix, "no suffix");
+
+  str = w_string_new_typed(".beginwithdot", W_STRING_UNICODE);
+  suffix = w_string_suffix(str);
+  expect_suffix = w_string_new_typed("beginwithdot", W_STRING_UNICODE);
+  ok(w_string_equal(suffix, expect_suffix), "begin with dot");
```
Comment (wez) [inline tests/string.cpp]: using `w_string` you can phrase this more clearly:

```
ok(w_string(".beginwithdot").suffix() == w_string("beginwithdot"), "begin with dot");
```

File: tests/string.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  ok(w_string_equal(suffix, expect_suffix), "java suffix");
+
+  // many '.' in name
+  str = w_string_new_typed("index.android.bundle", W_STRING_UNICODE);
+  suffix = w_string_suffix(str);
+  expect_suffix = w_string_new_typed("bundle", W_STRING_UNICODE);
+  ok(w_string_equal(suffix, expect_suffix), "multi dots suffix");
+
+  char too_long_name[129] = {0};
+  for (char &ch : too_long_name) {ch = 'a'; }
```
Comment (wez) [inline tests/string.cpp]: prefer something like this:

```
std::string longName(129, 'a');
longName[0] = '.';
str = w_string(longName.data(), longName.size());
suffix = str.suffix();
```

File: tests/string.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  too_long_name[0] = '.';
+  str = w_string_new_len_typed(too_long_name, sizeof(too_long_name), W_STRING_UNICODE);
+  suffix = w_string_suffix(str);
+  ok(!suffix, "too long suffix");
+
+  if (str) w_string_delref(str);
```
Comment (wez) [inline tests/string.cpp]: you won't need these delrefs when you switch to `w_string`.

File: tests/string.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  if (suffix) w_string_delref(suffix);
+  if (expect_suffix) w_string_delref(expect_suffix);
+
+}
+
+int main(int argc, char **argv)
```
Comment (wez) [inline tests/string.cpp]: since we're in c++, you can just omit the argument names of unused parameters:

```
int main(int, char**) {
```

File: tests/string.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+{
+  (void)argc;
```
Comment (wez) [inline tests/string.cpp]: and then you won't need these

File: winbuild/Makefile [hunk 9]
```diff
@@ -136,7 +136,8 @@ TEST_DIR_SRCS_CPP=\
 	tests\ht_test.cpp \
 	tests\ignore_test.cpp \
 	tests\log_stub.cpp \
-	tests\log.cpp
+	tests\log.cpp \
+	tests\string.cpp
 
 OBJS=$(SRCS_CPP:.cpp=.obj) $(SRCS_C:.c=.obj)
 TEST_OBJS=$(TEST_SRCS_CPP:.cpp=.obj) $(TEST_SRCS_C:.c=.obj)
```
Comment (wez) [inline winbuild/Makefile]: github won't let me comment on precisely the correct line, but you'll need to add `tests\string.exe` to the list of `TESTS=` that is a couple of lines below this in order for the windows build to build and run your new test

**Comment by wez** [comment]
FWIW, https://github.com/facebook/watchman/commit/7a9d768db0d964a60b34f649b82b4e8610e81f87 adds a string.cpp test case and the associated build plumbing, so you'll need to rebase your changes; the good news is that you can focus on adding the suffix test cases to that now existing file rather than worrying about making it build!

**Comment by kaitian521** [comment]
i see the commit you pick up above which save me a lot of time.

**Comment by kaitian521** [comment]
maybe I can rename all files that have suffix '.', so I think this issue can be closed.

File: Makefile.am [hunk 10]
```diff
@@ -134,8 +134,8 @@ TESTS = \
 		tests/bser.t \
 		tests/ignore.t \
 		tests/pending.t \
-		tests/string.t \
 		tests/log.t \
+		tests/string.t \
```
Comment (wez) [inline Makefile.am]: please revert the changes in this file; since you're now adding to the existing test, this pull request doesn't need to make changes here :-)

File: Makefile.am [hunk 11]
```diff
@@ -259,6 +259,7 @@ tests_log_t_SOURCES = \
 	tests/log.cpp \
 	log.cpp
 
+
```
Comment (wez) [inline Makefile.am]: same here

File: string.cpp [hunk 12]
```diff
@@ -624,6 +624,11 @@ w_string_t *w_string_suffix(w_string_t *str)
       }
       *buf = '\0';
       return w_string_new_typed(name_buf, str->type);
+    } else {
```
Comment (wez) [inline string.cpp]: prefer to phrase this block as:

```
} else if (str->len - end >= sizeof(name_buf)) {
   // We haven't found the '.' yet but the suffix will never fit in our local buffer
   return nullptr;
}
```

**Comment by kaitian521** [comment]
sorry for my poor english  😊
